src/pybind/mgr/cephadm/schedule.py

```diff
@@ -0,0 +1,183 @@
+    def place(self, host_pool, count=None):
+        # type: (List, Optional[int]) -> List[HostPlacementSpec]
+        if not host_pool:
+            return []
+        host_pool = [x for x in host_pool]
+        # shuffle for pseudo random selection
+        random.shuffle(host_pool)
+        return host_pool[:count]
+
+
+class HostAssignment(object):
+    """
+    A class to detect if hosts are being passed imperative or declarative
+    If the spec is populated via the `hosts/hosts` field it will not load
+    any hosts into the list.
+    If the spec isn't populated, i.e. when only num or label is present (declarative)
+    it will use the provided `get_host_func` to load it from the inventory.
+
+    Schedulers can be assigned to pick hosts from the pool.
+    """
+
+    def __init__(self,
+                 spec,  # type: ServiceSpec
+                 get_hosts_func,  # type: Callable[[Optional[str]],List[str]]
+                 get_daemons_func, # type: Callable[[str],List[orchestrator.DaemonDescription]]
+
+                 filter_new_host=None, # type: Optional[Callable[[str],bool]]
+                 scheduler=None,  # type: Optional[BaseScheduler]
+                 ):
+        assert spec and get_hosts_func and get_daemons_func
+        self.spec = spec  # type: ServiceSpec
+        self.scheduler = scheduler if scheduler else SimpleScheduler(self.spec.placement)
+        self.get_hosts_func = get_hosts_func
+        self.get_daemons_func = get_daemons_func
+        self.filter_new_host = filter_new_host
+        self.service_name = spec.service_name()
+
+
+    def validate(self):
+        self.spec.validate()
+
+        if self.spec.placement.hosts:
+            explicit_hostnames = {h.hostname for h in self.spec.placement.hosts}
+            unknown_hosts = explicit_hostnames.difference(set(self.get_hosts_func(None)))
+            if unknown_hosts:
+                raise OrchestratorValidationError(
+                    f'Cannot place {self.spec.one_line_str()} on {", ".join(sorted(unknown_hosts))}: Unknown hosts')
+
+        if self.spec.placement.host_pattern:
+            pattern_hostnames = self.spec.placement.pattern_matches_hosts(self.get_hosts_func(None))
+            if not pattern_hostnames:
+                raise OrchestratorValidationError(
+                    f'Cannot place {self.spec.one_line_str()}: No matching hosts')
+
+        if self.spec.placement.label:
+            label_hostnames = self.get_hosts_func(self.spec.placement.label)
+            if not label_hostnames:
+                raise OrchestratorValidationError(
+                    f'Cannot place {self.spec.one_line_str()}: No matching '
+                    f'hosts for label {self.spec.placement.label}')
+
+    def place(self):
+        # type: () -> List[HostPlacementSpec]
+        """
+        Load hosts into the spec.placement.hosts container.
+        """
+
+        self.validate()
+
+        count = self.spec.placement.count
+        assert count != 0
```

Comment:
Wouldn't that be rather something for `validate()`?

Reply:
placement.validate() will fail, if `count == 0`. We should never assert here.